perma-payments/perma_payments/models.py

```diff
@@ -0,0 +1,453 @@
+            #
+            # The subscription has been canceled.
+            ('Cancelled', 'Cancelled'),
+            # All payments have been processed (installments subscriptions).
+            # You see this status one or two days after the last payment is processed.
+            # (As of 8/16/17, should never be returned to Perma Payments, since we are not selling installment plans.)
+            ('Completed', 'Completed'),
+            # The subscription is active, and the payments are up to date.
+            ('Current', 'Current'),
+            # The subscription is on hold because all payment attempts have failed
+            # or a scheduled payment failed for a reason that requires your intervention.
+            # You can see the subscriptions on hold in the daily Payment Exception Report.
+            # For more information about holding profiles, see the Payment Tokenization documentation:
+            # http://apps.cybersource.com/library/documentation/dev_guides/Payment_Tokenization/html
+            ('Hold', 'Hold'),
+            # The subscription has been updated and a new subscription ID has been assigned to it.
+            # (As of 8/16/17, should never be returned to Perma Payments, since our account is not
+            # configured to use 16-digit format-preserving payment tokens.)
+            ('Superseded', 'Superseded')
+        )
+    )
+    status_updated = models.DateTimeField(auto_now=True)
+    cancellation_requested = models.BooleanField(
+        default=False
+    )
+
+
+    @classmethod
+    def registrar_standing_subscription(cls, registrar):
+        standing = cls.objects.filter(registrar=registrar, status__in=STANDING_STATUSES).order_by('id')
+        count = len(standing)
+        if count == 0:
+            return None
+        if count > 1:
+            logger.error("Registrar {} has multiple standing subscriptions ({})".format(registrar, count))
+            if settings.RAISE_IF_MULTIPLE_SUBSCRIPTIONS_FOUND:
+                raise cls.MultipleObjectsReturned
+        # In the extremely unlikely (incorrect!) condition that a registrar has multiple standing subscriptions,
+        # return the oldest. Probably, something went wrong with an update request;
+        # we should cancel the new subscription(s), use the original, and if needed update the original one.
+        return standing[0]
+
+
+    def can_be_altered(self):
+        return self.status in STANDING_STATUSES and not self.cancellation_requested
+
+
+    def update_status_after_cs_decision(self, decision, redacted_response):
+        decision_map = {
+            # Successful transaction. Reason codes 100 and 110.
+            'ACCEPT': {
+               'status': 'Current',
+               'log_level': logging.INFO,
+               'message': "Subscription request for registrar {} (subscription request {}) accepted.".format(self.registrar, self.subscription_request.pk)
```

Comment:
You've got a lot of strings sprinkled throughout your code. Maybe in future versions consider moving them into the constants.py (or some other) module? This will potentially help with scattered typos (they'll all now live in one file)

Reply:
Hmm do you mean the strings for log messages? I definitely want to avoid magic numbers/strings/undeclared constants, but I'm having trouble finding any. Would love to talk about this some more.

Re: log messages: do your instincts say it would be preferable to have a file that contained all possible logging messages, rather than building the logging message where you intend to use it? I'm not all that comfortable working with format strings: line 154, above, the one you commented on, is a format string relying on two variables. E.g., the logging message is customized, to include info about the particular registrar and subscription request it's logging about. I don't know a way of keeping format strings like that in a file like constants.py, ready to be customized here, when self.registrar and self.subscription_request.pk have meaning. Do you? Sounds like a neat thing to do!